### PR TITLE
[multi head attention] implement finalize @open sesame 8/22 14:00

### PR DESCRIPTION
**Pull request by lhs8928**
- implement multi head attention finalizeCommon, finalize
    
    Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

### CODE REVIEW

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 1]
```diff
@@ -77,10 +77,415 @@ enum AttentionParams {
...
+  if (context.getNumInputs() < 2 || context.getNumInputs() > 4) {
+    throw std::runtime_error("Multi head Attention layer needs 2-4 inputs.");
+  }
+
+  NNTR_THROW_IF(
+    std::get<props::ReferenceFramework>(multi_head_attention_props).empty(),
+    std::invalid_argument)
+    << "reference_framework property is not provided for layer "
+    << context.getName();
+  const Framework ref_framework =
```
Comment (jijoongmoon) [inline nntrainer/layers/multi_head_attention_layer.cpp:90]: I think defalut value of this property needs to be set.

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 2]
```diff
@@ -77,10 +77,415 @@ enum AttentionParams {
...
+  NNTR_THROW_IF(std::get<props::QueryDim>(multi_head_attention_props).empty(),
+                std::invalid_argument)
+    << "query_dim property is not provided for layer " << context.getName();
+  const unsigned int query_dim_prop =
+    std::get<props::QueryDim>(multi_head_attention_props).get();
+
+  if (std::get<props::KeyDim>(multi_head_attention_props).empty()) {
+    std::get<props::KeyDim>(multi_head_attention_props).set(query_dim_prop);
+  } else {
+    NNTR_THROW_IF(ref_framework == Framework::tensorflow, std::invalid_argument)
```
Comment (jijoongmoon) [inline nntrainer/layers/multi_head_attention_layer.cpp:108]: I think it might be better take one way to get the property as I describe https://github.com/nnstreamer/nntrainer/pull/1956/files#r926149987

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 3]
```diff
@@ -77,10 +77,415 @@ enum AttentionParams {
...
+  }
+  const unsigned int value_dim_prop =
+    std::get<props::ValueDim>(multi_head_attention_props).get();
+
+  if (std::get<props::ProjectedKeyDim>(multi_head_attention_props).empty()) {
+    if (ref_framework == Framework::pytorch) {
+      std::get<props::ProjectedKeyDim>(multi_head_attention_props)
+        .set(query_dim_prop / num_heads);
+    } else {
+      NNTR_THROW_IF(ref_framework == Framework::tensorflow,
```
Comment (jijoongmoon) [inline nntrainer/layers/multi_head_attention_layer.cpp:132]: same here.

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 1 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp]: fixed

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 2 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp]: As I commented in https://github.com/nnstreamer/nntrainer/pull/1956/files#r934121659, ProjectedkeyDim is not alias of KeyDim but as you recommended if the ProjectedKeyDim is not given then set the default value using the Keydim.

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 3 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp]: Done

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 4]
```diff
@@ -73,10 +73,317 @@ enum AttentionParams {
   transposed_attention_output,
 };
 
-void MultiHeadAttentionLayer::finalizeCommon(InitLayerContext &context) {}
+void MultiHeadAttentionLayer::finalizeCommon(InitLayerContext &context) {
```
Comment (jijoongmoon) [inline nntrainer/layers/multi_head_attention_layer.cpp:76]: Do we need `finalizeCommon` else where?

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 5]
```diff
@@ -73,10 +73,317 @@ enum AttentionParams {
...
+  if (context.getNumInputs() < 2 || context.getNumInputs() > 4) {
+    throw std::runtime_error("Multi head Attention layer needs 2-4 inputs.");
```
Comment (jijoongmoon) [inline nntrainer/layers/multi_head_attention_layer.cpp:78]: How about decribe more details.. what the inputs are.

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 6]
```diff
@@ -73,10 +73,317 @@ enum AttentionParams {
...
+  const bool provide_attention_mask =
+    std::get<props::ProvideAttentionMask>(multi_head_attention_props).get();
+
+  if (std::get<props::AverageAttentionWeight>(multi_head_attention_props)
+        .empty()) {
+    std::get<props::AverageAttentionWeight>(multi_head_attention_props)
+      .set(true);
+  }
+
+  inout_idx[INOUT_INDEX::QUERY] = INOUT_INDEX::QUERY;
```
Comment (jijoongmoon) [inline nntrainer/layers/multi_head_attention_layer.cpp:96]: Looks like inout_idx has same index. How about using just INOUT_INDEX::QUERY?

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 7]
```diff
@@ -73,10 +73,317 @@ enum AttentionParams {
...
+  auto &weight_initializer =
+    std::get<props::WeightInitializer>(*layer_impl_props).get();
+  auto &bias_initializer = std::get<props::BiasInitializer>(*layer_impl_props);
+  auto &weight_regularizer =
+    std::get<props::WeightRegularizer>(*layer_impl_props);
+  auto &weight_regularizer_constant =
+    std::get<props::WeightRegularizerConstant>(*layer_impl_props);
+  const float &weight_decay =
+    std::get<props::WeightDecay>(*layer_impl_props).get();
+  const float &bias_decay = std::get<props::BiasDecay>(*layer_impl_props).get();
```
Comment (jijoongmoon) [inline nntrainer/layers/multi_head_attention_layer.cpp:202]: How about move these layer_impl property to `finalizeCommon` if it should exist.

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 8]
```diff
@@ -73,10 +73,317 @@ enum AttentionParams {
...
+                                : empty_dim;
+
+  const unsigned int batch_size = query_dim.batch();
+  const unsigned int query_height = query_dim.height();
+  const unsigned int query_width = query_dim.width();
+  const unsigned int key_height = key_dim.height();
+  const unsigned int key_width = key_dim.width();
+  const unsigned int value_height = value_dim.height();
+  const unsigned int value_width = value_dim.width();
+
```
Comment (jijoongmoon) [inline nntrainer/layers/multi_head_attention_layer.cpp:237]: I think these are duplicated code with in `FinalizeCommon`

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 9]
```diff
@@ -73,10 +73,317 @@ enum AttentionParams {
...
+  TensorDim projected_value_dim(
+    {batch_size, 1, value_height, num_heads * projected_value_dim_prop});
+  weight_idx[AttentionParams::projected_value] = context.requestTensor(
+    projected_value_dim, "projected_value", Tensor::Initializer::NONE, true,
+    TensorLifespan::FORWARD_FUNC_LIFESPAN);
+  weight_idx[AttentionParams::d_projected_value] = context.requestTensor(
+    projected_value_dim, "d_projected_value", Tensor::Initializer::NONE, false,
+    TensorLifespan::BACKWARD_FUNC_LIFESPAN);
+
+  TensorDim transposed_projected_query_dim(
```
Comment (jijoongmoon) [inline nntrainer/layers/multi_head_attention_layer.cpp:313]: Transpose matrix could be removed. please check it.

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 10]
```diff
@@ -73,10 +73,317 @@ enum AttentionParams {
...
+    Tensor::Initializer::NONE, true, TensorLifespan::ITERATION_LIFESPAN);
+  TensorDim transposed_projected_value_dim(
+    {batch_size, num_heads, value_height, projected_value_dim_prop});
+  weight_idx[AttentionParams::transposed_projected_value] =
+    context.requestTensor(
+      transposed_projected_value_dim, "transposed_projected_value",
+      Tensor::Initializer::NONE, true, TensorLifespan::ITERATION_LIFESPAN);
+
+  TensorDim attention_score_dim(
+    {batch_size, num_heads, query_height, key_height});
```
Comment (jijoongmoon) [inline nntrainer/layers/multi_head_attention_layer.cpp:332]: please check if we could save the memory further.

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 6 above)
Comment (jijoongmoon) [reply nntrainer/layers/multi_head_attention_layer.cpp:96]: I missed the line #98.

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 10 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp:332]: Attention weight is calculated by apply softmax for attention score.
For now NNTrainer do not support inplace operation for softmax.
I will check if inplace operation for softmax is possible(I guess it is) and if it is I will make a another pr.

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 4 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp]: Nope. Removed finalizeCommon and merged it with finalize

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 5 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp]: Done

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 7 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp]: Done

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 8 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp]: Done

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 9 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp:270]: For now NNTrainer does not support inplace transpose operation.
First thing to do is check transpose is needed and if it is we need to implement inplace transpose operation

Comment (jijoongmoon) [reply nntrainer/layers/multi_head_attention_layer.cpp:270]: If the projected tensor can be overwritten, then it is better to use temporal tensor inside of transpose tensor op.

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 11]
```diff
@@ -73,10 +72,273 @@ enum AttentionParams {
   transposed_attention_output,
 };
 
-void MultiHeadAttentionLayer::finalizeCommon(InitLayerContext &context) {}
-
 void MultiHeadAttentionLayer::finalize(InitLayerContext &context) {
-  finalizeCommon(context);
+  if (context.getNumInputs() < 3 || context.getNumInputs() > 4) {
```
Comment (jijoongmoon) [inline nntrainer/layers/multi_head_attention_layer.cpp:76]: because the finalize function is quite long, how about add some comments which explain more details about the tensors and weights.

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 9 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp]: Done

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 11 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp:68]: Done

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 12]
```diff
@@ -55,28 +54,257 @@ enum AttentionParams {
...
   fc_bias,
   projected_query,
-  d_projected_query,
   projected_key,
-  d_projected_key,
   projected_value,
-  d_projected_value,
-  transposed_projected_query,
-  transposed_projected_key,
-  transposed_projected_value,
```
Comment (jijoongmoon) [inline nntrainer/layers/multi_head_attention_layer.cpp:65]: It is great! Now we can save more memory!

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 13]
```diff
@@ -55,28 +54,257 @@ enum AttentionParams {
...
   attention_score,
   d_attention_score,
-  attention_mask,
+  // attention_mask,
```
Comment (jihochu) [inline nntrainer/layers/multi_head_attention_layer.cpp:61]: need to remove? or leave comment.

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 14]
```diff
@@ -55,28 +54,257 @@ enum AttentionParams {
...
-  d_attention_output,
-  transposed_attention_output,
 };
 
-void MultiHeadAttentionLayer::finalizeCommon(InitLayerContext &context) {}
-
 void MultiHeadAttentionLayer::finalize(InitLayerContext &context) {
-  finalizeCommon(context);
+  if (context.getNumInputs() < 3 || context.getNumInputs() > 4) {
+    throw std::runtime_error("Multi head Attention layer needs 3 or 4 inputs. "
```
Comment (jihochu) [inline nntrainer/layers/multi_head_attention_layer.cpp:69]: invalid argument with NNTR_THROW_IF ?

File: nntrainer/layers/multi_head_attention_layer.cpp [hunk 15]
```diff
@@ -55,28 +54,257 @@ enum AttentionParams {
...
+  NNTR_THROW_IF(query_dim.channel() != 1, std::invalid_argument)
+    << "Dimension of input query channel: " << query_dim.channel()
+    << " is not 1 for layer " << context.getName();
+  NNTR_THROW_IF(key_dim.channel() != 1, std::invalid_argument)
+    << "Dimension of input key channel: " << key_dim.channel()
+    << " is not 1 for layer " << context.getName();
+  NNTR_THROW_IF(value_dim.channel() != 1, std::invalid_argument)
+    << "Dimension of input value channel: " << value_dim.channel()
+    << " is not 1 for layer " << context.getName();
+  NNTR_THROW_IF(provide_attention_mask && mask_dim.channel() != num_heads,
```
Comment (jihochu) [inline nntrainer/layers/multi_head_attention_layer.cpp:165]: How about extect ```provide_attention_mask``` to other throw ?
line 165, 175, 181 are using it.

**Review by DonghakPark — APPROVED** [review]
LGTM

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 13 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp:62]: Added a comment

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 14 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp]: Done

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 15 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp:166]: In multi head attention the input mask might not given so before checking the mask dimension checking the mask is provided or not is needed.
So I think we can't extract provide_attention_mask to other throw.

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 14 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.cpp]: I will make a another pr for other layers

**Review by songgot — APPROVED** [review]
LGTM

File: nntrainer/layers/multi_head_attention_layer.cpp (hunk 15 above)
Comment (jihochu) [reply nntrainer/layers/multi_head_attention_layer.cpp:166]: Oh, I see. thanks for checking that.

**Review by jihochu — APPROVED** [review]
LGTM
